src/axom/quest/MeshTester.hpp

```diff
@@ -33,12 +61,196 @@ enum class WatertightStatus : signed char
+  FloatType* ymin = axom::allocate< FloatType >( ncells );
+  FloatType* zmin = axom::allocate< FloatType >( ncells );
+
+  FloatType* xmax = axom::allocate< FloatType >( ncells );
+  FloatType* ymax = axom::allocate< FloatType >( ncells );
+  FloatType* zmax = axom::allocate< FloatType >( ncells );
+
+  // Each access-aligned bounding box represented by 2 (x,y,z) points
+  FloatType* aabbs = axom::allocate< FloatType >( ncells * stride );
+
+  // Finding degenerate indices and bounding boxes from Triangles
+  Triangle3 t1 = Triangle3();
+  Triangle3 t2 = Triangle3();
+
+  for (int i=0 ; i < ncells ; i++)
+  {
+    t1=getMeshTriangle(i, surface_mesh);
+    tris[i] = t1;
+
+    if (t1.degenerate())
+    {
+      degenerateIndices.push_back(i);
+    }
+
+    SpatialBoundingBox triBB = compute_bounds(t1);
+
+    const IndexType offset = i * stride;
+
+    xmin[i] = aabbs[ offset ]     = triBB.getMin()[0];
+    ymin[i] = aabbs[ offset + 1 ] = triBB.getMin()[1];
+    zmin[i] = aabbs[ offset + 2 ] = triBB.getMin()[2];
+
+    xmax[i] = aabbs[ offset + 3 ] = triBB.getMax()[0];
+    ymax[i] = aabbs[ offset + 4 ] = triBB.getMax()[1];
+    zmax[i] = aabbs[ offset + 5 ] = triBB.getMax()[2];
+  }
+
+  // Construct BVH
+  axom::spin::BVH< NDIMS, ExecSpace, FloatType > bvh( aabbs, ncells );
+  bvh.setScaleFactor( 1.0 ); // no scaling
+  bvh.build( );
+
+  // Run find algorithm
+  IndexType* offsets    = axom::allocate< IndexType >( ncells );
+  IndexType* counts     = axom::allocate< IndexType >( ncells );
+  IndexType* candidates = nullptr;
+  bvh.findBoundingBoxes( offsets, counts, candidates, ncells, xmin, xmax,
+                         ymin, ymax, zmin, zmax );
+
+  // Get the total number of candidates
+  int totalCandidates = 0;
+  for (int i = 0 ; i < ncells ; i++ )
+  {
+    totalCandidates += counts[i];
+  }
+
+  //Deallocate no longer needed variables
+  axom::deallocate(aabbs);
+
+  axom::deallocate(xmin);
+  axom::deallocate(ymin);
+  axom::deallocate(zmin);
+
+  axom::deallocate(xmax);
+  axom::deallocate(ymax);
+  axom::deallocate(zmax);
+
+  int* intersection_pairs =
+    axom::allocate <int> (totalCandidates * 2);
+
+  using ATOMIC_POL =
+          typename axom::execution_space< ExecSpace >::atomic_policy;
+  int* counter = axom::allocate <int> (1);
+  counter[0] = 0;
+
+  for_all< ExecSpace >( ncells, AXOM_LAMBDA (IndexType i)
+  {
+    for (int j = 0 ; j < counts[i] ; j++)
+    {
+      int candidate_index = candidates[ offsets[i] + j];
+      if (i != candidate_index && i < candidate_index)
+      {
+        if (primal::intersect(tris[i], tris[candidate_index]))
+        {
+          auto idx = RAJA::atomicAdd<ATOMIC_POL>(counter, 2);
+          intersection_pairs[idx] = i;
+          intersection_pairs[idx + 1] = candidate_index;
+        }
+      }
+    }
+  } );
+
+  // Initialize pairs of clashes
+  for (int i = 0 ; i < counter[0] ; i += 2)
+  {
+    intersections.push_back(std::make_pair(intersection_pairs[i],
```

Comment:
I think I'm missing something -- when does the memory associated with `intersection_pairs` get moved from device to host (for the `raja-cuda` runs)?